netconn.go

```diff
@@ -151,16 +186,18 @@ func (c *netConn) SetWriteDeadline(t time.Time) error {
 	if t.IsZero() {
 		c.writeTimer.Stop()
 	} else {
-		c.writeTimer.Reset(t.Sub(time.Now()))
+		c.writeTimer.Reset(time.Until(t))
 	}
+	c.afterWriteDeadline.Store(false)
 	return nil
 }
 
 func (c *netConn) SetReadDeadline(t time.Time) error {
 	if t.IsZero() {
 		c.readTimer.Stop()
 	} else {
-		c.readTimer.Reset(t.Sub(time.Now()))
+		c.readTimer.Reset(time.Until(t))
 	}
+	c.afterReadDeadline.Store(false)
```

Comment:
while highly unlikely this can race with the timer routine as it's possible the timer has already set off and the timer goroutine has already set `afterReadDeadline` to true before this goroutine hits this line.